### DISCUSSION TITLE
PNG/other image support on login screen

### DISCUSSION THREAD

**Discussion by matt335672** [neutrinolabs/xrdp#1931] [Ideas]
Hi all,

This is related to #1929.

Having had a look at the code linked in that PR, it's a relatively simple matter to get the login screen to optionally support resizeable PNG images in addition to the existing fixed .bmp images. The code provides some crude resizing support and allows an image to be alpha-blended with a single background colour. There's some work required, like adding hooks to autoconf for libpng, and adding a few controls for scaling, etc to xrdp.ini. That could all be in place for the next release.

This would allow organisations or distros to more easily support customisation of the xrdp login screen.

PNG support is available on pretty much every platform out there, and this is unlikely to change very soon.

I've researched other options briefly which would offer more flexibility but need a lot more work.

   - STB mentioned in https://stackoverflow.com/questions/50079/c-image-library. Advantages; no additional libraries required. Disadvantages; may impose extra maintenance overhead on us in the future; increased executable size.
   - [imlib2](https://sourceforge.net/projects/enlightenment/files/imlib2-src/). Advantages; seems to be available on most Linuxes and FreeBSD.
   - anything else?

What do people think? I'm personally tempted to go for the quick win, given all the other work we've got on at the moment.

**Comment by simdev5281** [comment]
For my company it would allow us to continue to use a corporate background styling without the penalty of a large image being required to be downloaded over the limited VPN connection first. I think its sensible to offer more than just bitmap as a standard for the images in any case with the added transparency element of PNG allowing for slicker integration into the login window.

**Reply by metalefty** [reply]
The BMP image is not transferred as-is. Such a penalty doesn't exist in the first place.

**Reply by simdev5281** [reply]
Really there's definitely some form of penalty. We have an image 4000x2000 in bitmap form is 30MB and then i reduced to 1080x540 at around 5MB.

The difference in time was instant versus around 5 to 10 seconds to render. Where is the delay coming from if its not size or data related? Obviously I don't know the intricacies of the program so its probably confirmation bias but i'd be interested to know what the solution is. I've tried reducing the format bitmap quality below 24bit but just end with a black background being rendered.

**Reply by matt335672** [reply]
I think I can see what's happening here.

The patched code will resize the loaded image to the screen size before sending it, hence it will be quicker. I think the unpatched code is sending the full-size bitmap, most of which is being discarded by the client on receipt.

Another option which should probably be included (but isn't) is for the login screen to honour the experience setting in the mstsc.exe client:-

![desktop](https://user-images.githubusercontent.com/30179339/123241080-c0f15d00-d4d8-11eb-88df-5cd0d592cf55.png)

It seems the patched code doesn't currently honour that setting.

**Reply by metalefty** [reply]
The screen image is encoded to RDP codec and transferred from server to client. The direct transfer of BMP images is not happening. Likewise, even xrdp support PNG images for login screen, it doesn't mean the actual transferred data is compressed by PNG.

PNG support on login screen doesn't reduce RDP traffic at all. They're separate issues.

**Reply by simdev5281** [reply]
@matt335672 that would explain it, so i've confused things by assuming it was difference between BMP and PNG. So what I'm really asking for is scaling support.

**Reply by matt335672** [reply]
@simdev5281 - I suspect your corporate backdrop is one that looks similar at different zoom levels. So you wouldn't necessarily notice whether it was being zoomed or not.

The patch you've linked above has got some really crude scaling support - it just does single-pixel sampling. So that will explain why you're using a large image in the first place. Any decent image library will implement smooth scaling which will not only produce a better result, but might mean you can distribute smaller images in the first place.

**Comment by metalefty** [comment]
I +1 to imlib2.

**Reply by matt335672** [reply]
I'm inclined to agree, despite the extra work. An important consideration here is that xrdp runs as root on most platforms.

imlib2 is a mature library, and being dynamically linked will get security patches if required. STB looks attractive, but will be statically linked which is less good from a security perspective.

**Reply by metalefty** [reply]
xrdp can be run as user privileges as Debian already does.  Only xrdp-sesman actually requires root privileges.

It is another discussion but it is a good chance to discuss whether we officially recommend running xrdp daemon in user privilege.

**Reply by matt335672** [reply]
I think it's a great idea.

I've written up how Debian does it on the [wiki](/neutrinolabs/xrdp/wiki/Running-the-xrdp-process-as-non-root). It's a bit fiddly, but we could bring quite a bit of it into autoconf I think which would make it more attractive to other systemd-based distros.

**Comment by matt335672** [comment]
This functionality is implemented in PR #1962, and as of 2021-09-01 is merged into devel.
